### PR TITLE
[Tensor] Remove calcGrad step for non-trainable layer

### PR DESCRIPTION
**Pull request by jihochu**
This patch is for implementing tarinable property behavior.

If a layer is set as non-trainable, it doesnot need to execute calcGrad step, so we can remove it from execution order, also skip gradient calculation.

Signed-off-by: Jiho Chu <jiho.chu@samsung.com>

### CODE REVIEW

**Comment by jihochu** [comment]
This PR is pending until @baek2sm 's  non-trainable layer test.

**Comment by jihochu** [comment]
It passed #2077

File: nntrainer/graph/network_graph.cpp [hunk 1]
```diff
@@ -85,14 +85,26 @@ int NetworkGraph::compile(const std::string &loss_type) {
...
   auto max_count = graph.size() * 3;
-  /** @todo: remove backwarding count for non-trainble layers */
+  size_t not_trainables = 0;
+
+  for (auto iter = cbegin(); iter != cend(); iter++) {
+    if (!(*iter)->getTrainable())
+      not_trainables++;
+  }
+
   for (auto iter = cbegin(); iter != cend(); iter++) {
```
Comment (lhs8928) [inline nntrainer/graph/network_graph.cpp:95]: How about we iterate from crbegin to crend? IMO iterate from end to begin will make logic more simple.
Please check https://gist.github.com/lhs8928/5610124f0de7787d1eb1b7248159f864

File: nntrainer/tensor/manager.cpp [hunk 2]
```diff
@@ -355,9 +355,8 @@ std::vector<Weight *> Manager::requestWeights(
     node.getExecutionOrder();
 
   std::vector<unsigned int> var_exec_order(
-    {forwarding_order, calcGradient_order, calcDerivative_order});
-  std::vector<unsigned int> default_grad_exec_order(
-    {calcGradient_order, calcDerivative_order});
+    {forwarding_order, calcDerivative_order});
```
Comment (lhs8928) [inline nntrainer/tensor/manager.cpp:358]: Can you please describe why did you remove calcGradient_order from var_exec_order?
I think weights are also needed in some cases when calculate gradient.
Please check https://github.com/nnstreamer/nntrainer/blob/main/nntrainer/layers/grucell.cpp#L472

File: nntrainer/tensor/manager.cpp [hunk 3]
```diff
@@ -369,6 +368,10 @@ std::vector<Weight *> Manager::requestWeights(
     auto &[dim, t_initializer, w_reg, w_reg_const, decay, clip_by_global_norm,
            need_gradient, name] = weights_spec.at(i);
     auto grad_exec_order = default_grad_exec_order;
+
+    if (trainable)
```
Comment (lhs8928) [inline nntrainer/tensor/manager.cpp:372]: This is trivial but I think grad_exec_order already used only the layer is trainable so this change will not affect the weights lifespan.

File: nntrainer/graph/network_graph.cpp [hunk 4]
```diff
@@ -85,14 +85,26 @@ int NetworkGraph::compile(const std::string &loss_type) {
...
-    auto calc_gradient_order = max_count - ((order_idx + 1) * 2);
+    auto calc_gradient_order =
+      max_count - ((order_idx + 1) * 2) - not_trainables;
     /** calc derivative is called right after calc_gradient */
     auto calc_derivative_order = calc_gradient_order + 1;
+    if (!node->getTrainable()) {
+      not_trainables--;
+      calc_gradient_order++;
+      calc_derivative_order = calc_gradient_order;
+    }
```
Comment (jijoongmoon) [inline nntrainer/graph/network_graph.cpp:107]: if we set ```calc_derivative_order = calc_gradient_order```, we might needs more test cases to confirm.

File: nntrainer/graph/network_graph.cpp [hunk 5]
```diff
@@ -85,14 +85,26 @@ int NetworkGraph::compile(const std::string &loss_type) {
 
 void NetworkGraph::setExecutionOrder() {
   auto max_count = graph.size() * 3;
-  /** @todo: remove backwarding count for non-trainble layers */
+  size_t not_trainables = 0;
```
Comment (jijoongmoon) [inline nntrainer/graph/network_graph.cpp:88]: I think, We need to check if the max_count is the same with ```std::get<2>(backward_iter_end->getExecutionOrder()``` . Because the max is used at ```tensor_manager->allocateTensors(std::get<2>(backward_iter_end->getExecutionOrder())```

File: nntrainer/tensor/manager.cpp (hunk 3 above)
Comment (lhs8928) [reply nntrainer/tensor/manager.cpp]: Yeah you're right.
I was confused.

**Review by lhs8928 — APPROVED** [review]
LGTM

File: nntrainer/graph/network_graph.cpp (hunk 5 above)
Comment (jihochu) [reply nntrainer/graph/network_graph.cpp]: it's changed to iterate reversely with @lhs8928 's comment.  please check.

**Review by jijoongmoon — APPROVED** [review]
LGTM
